README.md
This version is compiled (GCC 8.3) and tested with Raspbian Stretch (armv7) and Ubuntu Mate (aarch64).

Requirements:
 - guichan 0.8.1
 - libpng16
 - libSDL 1.2
 - libSDL_image
 - libSDL_ttf
 - libmpeg2
 - libmpg123
 - libxml2
 - libFLAC
 - capsimg from https://github.com/FrodeSolheim/capsimg

Install the libs to run emulator on Raspberry Pi:

      sudo apt-get install libsdl1.2debian
      sudo apt-get install libsdl-image1.2
      sudo apt-get install libsdl-gfx1.2
      sudo apt-get install libsdl-ttf2.0
      sudo apt-get install libmpg123
      sudo apt-get install libguichan-sdl
      sudo apt-get install libxml2
      sudo apt-get install mpeg2dec

Install the packages to compile uae4arm on Raspberry Pi:

      sudo apt-get install libsdl-dev
      sudo apt-get install libguichan-dev
      sudo apt-get install libsdl-ttf2.0-dev
      sudo apt-get install libsdl-gfx1.2-dev
      sudo apt-get install libxml2-dev
      sudo apt-get install libflac-dev
      sudo apt-get install libmpg123-dev
      sudo apt-get install libmpeg2-4-dev

Aarch64:

To run on Ubuntu mate, dispmanx is required:

      git clone https://github.com/raspberrypi/userland.git
      cd userland
      buildme --aarch64

Use dsp as audio driver:

      export SDL_AUDIODRIVER=dsp


Read changelog.txt for the history of development of UAE4ARM for Pandora. 

Credits
 - Toni Wilen (WinUAE)
 - Frode Solheim (FS-UAE)
 - Chui (first version of uae4all)
 - Pickle (initial Pandora port of uae4all)
 - AnotherGuest (uae4all 2.0, based on Symbian sources of uae4all)
 - john4p (maintainer of uae4all 2.0 for Pandora)
 - Lyubomyr Lisen (Android port, bug reports, new GUI for uae4all2)
 - Jens Heitmann (idea and first versions of JIT in ARAnyM project)
 - Chips (Raspberry Pi port of uae4arm)
 - Midwan (Amiberry)
 - There are some more people who added code, but it’s hard to get everyone…
 


changelog.txt
2021-04-02 - For better performance, let compiler decide about inlining (profiler can also do a better job with this)
             JIT: one more host register (LR) for 32 bit version
             JIT: reduce usage of scratch register
2021-03-21 - Bugfix: use hardfile with A600/A1200/A4000 IDE controller
2021-03-20 - Fast copper improvement
           - Small bugfixes and cleanups
2021-03-16 - Bugfix JIT AARCH64 (DIVS)
2021-03-16 - Fixes for load/save of statefiles
2021-03-14 - Fixes for fast copper
2021-03-13 - Merge with WinUAE up to 4.5.0 b17
2021-03-07 - Two small bugfixes (rp9 parsing; fmv when line doubling)
2021-03-06 - Merge WinUAE 4.4.0
           - Added cycle exact emulation for 68000 cpu (slows down also regular emulation a bit)
           - Added drivesound
           - Added disk info dialog
           - Added hardware info panel
           - Added support for VHD harddrive images
           - Added pause emualtion
           - Added turbo mode (warp mode)
           - Fixed input mapping via dialog
           - Added emulation of serial port for games/demos using this as timing source (no real serial in/out)
           - Added on screen message when insert/remove disk or CD
2020-06-27 - Merge WinUAE 4.3.0
           - Added scsi.device (use CD in A1200/A4000 emulation)
           - Support für 32 bit native display
           - Use GCC 9.2
           - Support for Raspberry Pi 4 (64 bit version much faster than 32 bit)
2020-02-21 - JIT: implemented BFINS
2020-02-17 - JIT: bugfix CMP.W (32 bit only)
2020-02-14 - JIT: bugfix shift opcodes
2020-01-30 - JIT: reduced size of emitted code (30% smaller), slightly better performance
2020-01-28 - Bugfix JIT: FPU (32 bit), ORSR/EORSR/ANDSR, LINK/UNLK
2019-08-17 - Undone some changes from WinUAE which causes failures
2019-08-15 - Merge with current WinUAE (2019-08-14)
           - Bugfix JIT (armv7 and aarch64)
2019-08-11 - Performance enhancement for JIT (armv7 and aarch64)
           - Cleanup after profiling
           - LDR/STR handling for aarch64 in sigsev_handler
2019-08-04 - Cleanup in JIT
2019-07-17 - Fixed sigsegv-handler for aarch64
2019-07-11 - Optimize aarch64 JIT
2019-07-10 - Bugfixes for aarch64 JIT
2019-07-09 - Added FPU JIT for aarch64 (needs more testing)
2019-06-07 - JIT for aarch64: minor improvements and fixes
2019-06-05 - JIT for aarch64 (without FPU)
2019-05-31 - aarch64 JIT: first opcodes working
2019-05-31 - Fix atomic functions for aarch64
2019-05-22 - Merge WinUAE 4.2.1
2019-05-20 - Preparations for aarch64 JIT
2019-05-18 - aarch64: fix 32bpp in Picasso, reduced size of some vars
2019-05-17   First version for aarch64
 - Tested on Raspberry Pi 3B with Ubuntu Mate 64Bit.
 - Compiled native on Raspberry (GCC 7.4.0) or using cross-compiler on Ubuntu PC (GCC 8.3.0).
 - There is still work to do:
    * No JIT (rewrite of code generator needed, because of different instruction set and instruction encoding)
    * Fix for atomic functions (get segfaults when using GCC builtins)
    * Optimize for aarch64
2019-04-12 - Merged WinUAE 4.2.0
2019-03-29 - Merged WinUAE 4.2.0 beta 6
2019-03-09 - Another fast copper update
2019-03-08 - Fast copper update (bit slower but more compatible)
           - CPU update from WinUAE 4.2.0 beta 4
2019-03-05 - Merged WinUAE 4.2.0 beta 4
2019-03-04 - Removed some unused stuff
2019-01-01 - Enabled CAPS/IPF support using capsimg from FS-UAE
2019-01-01 - Bugfix: select disk file in GUI
2018-12-08 - Merged WinUAE 4.1.0
2018-12-07 - Finished merge 4.0.1
2018-12-07 - Fixed capslock sync, updated statusline
2018-12-03 - JIT: Bugfix CMPA.W (fixes OS3.9 installer)
           - Fix for add folder as HD
2018-12-02 - Added some options in GUI and small fixes
2018-12-01 - Start merge with WinUAE 4.0.1
           - Rewrite of vsync (SDL1)
           - Now using input mapping like WinUAE. No more target specific options for input.
           - Removed some unused stuff.
2018-10-06 - JIT: Bugfix for issue #1, MVMEL #<data>.W,(An)+
2018-10-05 - Minor fixes, build with GCC 8.2 ok
2018-10-04 - JIT: Bugfix ROXR
2018-04-01 - JIT: performance enhancements, implemented ROXR
2018-03-09 - JIT: bugfixes and small enhancements
2018-02-28 - Fixes for non ARMv6T2 target
2018-02-23 - Small enhancements and fixes (for non ARMv6T2)
2018-02-16 - Performance enhancements in JIT and fix for MOVE16 in JIT
2018-02-14 - Implemented ROXL and DIVL (32 bit) for JIT
2018-02-08 - Bugfixes for FPU JIT
2018-01-29 - Added JIT for FPU
2018-01-20 - Merge with WinUAE 3.6.0
2017-11-29 - Added line doubling to remove flicker in interlace mode
2017-11-28 - Fixed problem when Amiga framerate differs from Raspi framerate

2017-11-20 - Bug fix for game War zone

2017-11-19 - Alpha version
   - Now using more code from WinUAE in custom.cpp
   - Minor changes in cpu emulation
   
2017-11-17 - Alpha version
   - Use sprite drawing code from WinUAE 3.5.0
   - Cleanup code
   - Bufix for render thread
   - Minor bug fixes
   
2017-11-08 - First version for Raspi with render thread
   - My first version for RaspberryPi using code from Chips and Midwan
   - I introduced new render thread for better performance on hosts with multi core cpus
   
2017-09-09 - Release version
   - Fixed frameskip (was broken in last two beta versions)
   - Fixed Picasso framerate (Pandora switches now to 60 Hz like the Amiga side)
   - Added help page for every panel in GUI (Help button or F1)
   - Fixed bug when emulator crashes or hangs after switching configs

2017-09-06 - Beta version
   - Merged lot of changes from WinUAE 3.5.0
   - Fixed possible segfault and bus error on two locations
   - Better performance by using gcc 7.2 with profiler 

2017-08-08 - Beta version
   - Fixed loading new config (now using same method like WinUAE)
   - Display message instead of crash in some situations
   - More cleanup of threads/pipes/mutex

2017-08-06 - Beta version
   - Merged lot of changes from WinUAE 3.3.0
   - Added panel "Quickstart"

src/crc32.cpp

#include "sysconfig.h"
#include "sysdeps.h"

#include "crc32.h"

static unsigned long crc_table32[256];
static unsigned short crc_table16[256];
static void make_crc_table (void)
{
    unsigned long c;
    unsigned short w;
    int n, k;
    for (n = 0; n < 256; n++) {
	c = (unsigned long)n;
	w = n << 8;
	for (k = 0; k < 8; k++) {
	    c = (c >> 1) ^ (c & 1 ? 0xedb88320 : 0);
	    w = (w << 1) ^ ((w & 0x8000) ? 0x1021 : 0);
	}
	crc_table32[n] = c;
	crc_table16[n] = w;
    }
}
uae_u32 get_crc32_val (uae_u8 v, uae_u32 crc)
{
    if (!crc_table32[1])
	make_crc_table();
    crc ^= 0xffffffff;
    crc = crc_table32[(crc ^ v) & 0xff] ^ (crc >> 8);
    return crc ^ 0xffffffff;
}
uae_u32 get_crc32 (uae_u8 *buf, int len)
{
    uae_u32 crc;
    if (!crc_table32[1])
	make_crc_table();
    crc = 0xffffffff;
    while (len-- > 0)
	crc = crc_table32[(crc ^ (*buf++)) & 0xff] ^ (crc >> 8);    
    return crc ^ 0xffffffff;
}
uae_u16 get_crc16( uae_u8 *buf, int len)
{
    uae_u16 crc;
    if (!crc_table32[1])
	make_crc_table();
    crc = 0xffff;
    while (len-- > 0)
	crc = (crc << 8) ^ crc_table16[((crc >> 8) ^ (*buf++)) & 0xff];
    return crc;
}

#ifndef GET_UINT32_BE
#define GET_UINT32_BE(n,b,i)                            \
{                                                       \
    (n) = ( (unsigned long) (b)[(i)    ] << 24 )        \
        | ( (unsigned long) (b)[(i) + 1] << 16 )        \
        | ( (unsigned long) (b)[(i) + 2] <<  8 )        \
        | ( (unsigned long) (b)[(i) + 3]       );       \
}
#endif
#ifndef PUT_UINT32_BE
#define PUT_UINT32_BE(n,b,i)                            \
{                                                       \
    (b)[(i)    ] = (unsigned char) ( (n) >> 24 );       \
    (b)[(i) + 1] = (unsigned char) ( (n) >> 16 );       \
    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
}
#endif

typedef struct
{
    unsigned long total[2];     /*!< number of bytes processed  */
    unsigned long state[5];     /*!< intermediate digest state  */
    unsigned char buffer[64];   /*!< data block being processed */
}
sha1_context;

static void sha1_starts( sha1_context *ctx )
{
    ctx->total[0] = 0;
    ctx->total[1] = 0;

    ctx->state[0] = 0x67452301;
    ctx->state[1] = 0xEFCDAB89;
    ctx->state[2] = 0x98BADCFE;
    ctx->state[3] = 0x10325476;
    ctx->state[4] = 0xC3D2E1F0;
}

static void sha1_process( sha1_context *ctx, unsigned char data[64] )
{
    unsigned long temp, W[16], A, B, C, D, E;

    GET_UINT32_BE( W[0],  data,  0 );
    GET_UINT32_BE( W[1],  data,  4 );
    GET_UINT32_BE( W[2],  data,  8 );
    GET_UINT32_BE( W[3],  data, 12 );
    GET_UINT32_BE( W[4],  data, 16 );
    GET_UINT32_BE( W[5],  data, 20 );
    GET_UINT32_BE( W[6],  data, 24 );
    GET_UINT32_BE( W[7],  data, 28 );
    GET_UINT32_BE( W[8],  data, 32 );
    GET_UINT32_BE( W[9],  data, 36 );
    GET_UINT32_BE( W[10], data, 40 );
    GET_UINT32_BE( W[11], data, 44 );
    GET_UINT32_BE( W[12], data, 48 );
    GET_UINT32_BE( W[13], data, 52 );
    GET_UINT32_BE( W[14], data, 56 );
    GET_UINT32_BE( W[15], data, 60 );

#define S(x,n) ((x << n) | ((x & 0xFFFFFFFF) >> (32 - n)))

#define R(t)                                            \
(                                                       \
    temp = W[(t -  3) & 0x0F] ^ W[(t - 8) & 0x0F] ^     \
           W[(t - 14) & 0x0F] ^ W[ t      & 0x0F],      \
    ( W[t & 0x0F] = S(temp,1) )                         \
)

#define P(a,b,c,d,e,x)                                  \
{                                                       \
    e += S(a,5) + F(b,c,d) + K + x; b = S(b,30);        \
}

    A = ctx->state[0];
    B = ctx->state[1];
    C = ctx->state[2];
    D = ctx->state[3];
    E = ctx->state[4];

#define F(x,y,z) (z ^ (x & (y ^ z)))
#define K 0x5A827999

    P( A, B, C, D, E, W[0]  );
    P( E, A, B, C, D, W[1]  );
    P( D, E, A, B, C, W[2]  );
    P( C, D, E, A, B, W[3]  );
    P( B, C, D, E, A, W[4]  );
    P( A, B, C, D, E, W[5]  );
    P( E, A, B, C, D, W[6]  );
    P( D, E, A, B, C, W[7]  );
    P( C, D, E, A, B, W[8]  );
    P( B, C, D, E, A, W[9]  );
    P( A, B, C, D, E, W[10] );
    P( E, A, B, C, D, W[11] );
    P( D, E, A, B, C, W[12] );
    P( C, D, E, A, B, W[13] );
    P( B, C, D, E, A, W[14] );
    P( A, B, C, D, E, W[15] );
    P( E, A, B, C, D, R(16) );
    P( D, E, A, B, C, R(17) );
    P( C, D, E, A, B, R(18) );
    P( B, C, D, E, A, R(19) );

#undef K
#undef F

#define F(x,y,z) (x ^ y ^ z)
#define K 0x6ED9EBA1

    P( A, B, C, D, E, R(20) );
    P( E, A, B, C, D, R(21) );
    P( D, E, A, B, C, R(22) );
    P( C, D, E, A, B, R(23) );
    P( B, C, D, E, A, R(24) );
    P( A, B, C, D, E, R(25) );
    P( E, A, B, C, D, R(26) );
    P( D, E, A, B, C, R(27) );
    P( C, D, E, A, B, R(28) );
    P( B, C, D, E, A, R(29) );
    P( A, B, C, D, E, R(30) );
    P( E, A, B, C, D, R(31) );
    P( D, E, A, B, C, R(32) );
    P( C, D, E, A, B, R(33) );
    P( B, C, D, E, A, R(34) );
    P( A, B, C, D, E, R(35) );
    P( E, A, B, C, D, R(36) );
    P( D, E, A, B, C, R(37) );
    P( C, D, E, A, B, R(38) );
    P( B, C, D, E, A, R(39) );

#undef K
#undef F

#define F(x,y,z) ((x & y) | (z & (x | y)))
#define K 0x8F1BBCDC

    P( A, B, C, D, E, R(40) );
    P( E, A, B, C, D, R(41) );
    P( D, E, A, B, C, R(42) );
    P( C, D, E, A, B, R(43) );
    P( B, C, D, E, A, R(44) );
    P( A, B, C, D, E, R(45) );
    P( E, A, B, C, D, R(46) );
    P( D, E, A, B, C, R(47) );
    P( C, D, E, A, B, R(48) );
    P( B, C, D, E, A, R(49) );
    P( A, B, C, D, E, R(50) );
    P( E, A, B, C, D, R(51) );
    P( D, E, A, B, C, R(52) );
    P( C, D, E, A, B, R(53) );
    P( B, C, D, E, A, R(54) );
    P( A, B, C, D, E, R(55) );
    P( E, A, B, C, D, R(56) );
    P( D, E, A, B, C, R(57) );
    P( C, D, E, A, B, R(58) );
    P( B, C, D, E, A, R(59) );

#undef K
#undef F

#define F(x,y,z) (x ^ y ^ z)
#define K 0xCA62C1D6

    P( A, B, C, D, E, R(60) );
    P( E, A, B, C, D, R(61) );
    P( D, E, A, B, C, R(62) );
    P( C, D, E, A, B, R(63) );
    P( B, C, D, E, A, R(64) );
    P( A, B, C, D, E, R(65) );
    P( E, A, B, C, D, R(66) );
    P( D, E, A, B, C, R(67) );
    P( C, D, E, A, B, R(68) );
    P( B, C, D, E, A, R(69) );
    P( A, B, C, D, E, R(70) );
    P( E, A, B, C, D, R(71) );
    P( D, E, A, B, C, R(72) );
    P( C, D, E, A, B, R(73) );
    P( B, C, D, E, A, R(74) );
    P( A, B, C, D, E, R(75) );
    P( E, A, B, C, D, R(76) );
    P( D, E, A, B, C, R(77) );
    P( C, D, E, A, B, R(78) );
    P( B, C, D, E, A, R(79) );

#undef K
#undef F

    ctx->state[0] += A;
    ctx->state[1] += B;
    ctx->state[2] += C;
    ctx->state[3] += D;
    ctx->state[4] += E;
}

/*
 * SHA-1 process buffer
 */
static void sha1_update( sha1_context *ctx, unsigned char *input, int ilen )
{
    int fill;
    unsigned long left;

    if( ilen <= 0 )
        return;

    left = ctx->total[0] & 0x3F;
    fill = 64 - left;

    ctx->total[0] += ilen;
    ctx->total[0] &= 0xFFFFFFFF;

    if( ctx->total[0] < (unsigned long) ilen )
        ctx->total[1]++;

    if( left && ilen >= fill )
    {
        memcpy( (void *) (ctx->buffer + left),
                (void *) input, fill );
        sha1_process( ctx, ctx->buffer );
        input += fill;
        ilen  -= fill;
        left = 0;
    }

    while( ilen >= 64 )
    {
        sha1_process( ctx, input );
        input += 64;
        ilen  -= 64;
    }

    if( ilen > 0 )
    {
        memcpy( (void *) (ctx->buffer + left),
                (void *) input, ilen );
    }
}

static const unsigned char sha1_padding[64] =
{
 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/*
 * SHA-1 final digest
 */
static void sha1_finish( sha1_context *ctx, unsigned char output[20] )
{
    unsigned long last, padn;
    unsigned long high, low;
    unsigned char msglen[8];

    high = ( ctx->total[0] >> 29 )
         | ( ctx->total[1] <<  3 );
    low  = ( ctx->total[0] <<  3 );

    PUT_UINT32_BE( high, msglen, 0 );
    PUT_UINT32_BE( low,  msglen, 4 );

    last = ctx->total[0] & 0x3F;
    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );

    sha1_update( ctx, (unsigned char *) sha1_padding, padn );
    sha1_update( ctx, msglen, 8 );

    PUT_UINT32_BE( ctx->state[0], output,  0 );
    PUT_UINT32_BE( ctx->state[1], output,  4 );
    PUT_UINT32_BE( ctx->state[2], output,  8 );
    PUT_UINT32_BE( ctx->state[3], output, 12 );
    PUT_UINT32_BE( ctx->state[4], output, 16 );
}

void get_sha1(uae_u8 *input, int len, uae_u8 *out)
{
    sha1_context ctx;

    sha1_starts( &ctx );
    sha1_update( &ctx, input, len );
    sha1_finish( &ctx, out );
}
char *get_sha1_txt(uae_u8 *input, int len)
{
    static char outtxt[SHA1_SIZE * 2 + 1];
    uae_u8 out[SHA1_SIZE];
    int i;
    char *p;

    p = outtxt;
    get_sha1(input, len, out);
    for (i = 0; i < SHA1_SIZE; i++) {
	sprintf(p, "%02X", out[i]);
	p += 2;
    }
    *p = 0;
    return outtxt;
}
